tests/launcher.py

```diff
@@ -478,7 +539,139 @@ def make_mesh(self):
+
+            stdout = open(out, 'w')
+            stderr = open(err, 'w')
+            c = subprocess.Popen(eosdcmd, stdout=stdout, stderr=stderr)
+            with pidf.open('w') as pidout:
+                pidout.write(str(c.pid))
+            try:
+                err_sl.unlink()
+            except FileNotFoundError:
+                pass
+            err_sl.symlink_to(err.name)
+        else:
+            Utils.Print(f'unstarted node command: {" ".join(eosdcmd)}')
+
+            with open(instance.data_dir_name / 'start.cmd', 'w') as f:
+                f.write(' '.join(eosdcmd))
+
+    def bounce(self, nodeNumbers):
+        for num in nodeNumbers:
+            for node in self.network.nodes.values():
+                if self.network.name + num == node.name:
+                    with open(node.data_dir_name / f'{Utils.EosServerName}.pid', 'r') as f:
+                        pid = int(f.readline())
+                        self.terminate_wait_pid(pid, raise_if_missing=False)
+                    self.launch(node)
+
+    def down(self, nodeNumbers):
+        for num in nodeNumbers:
+            for node in self.network.nodes.values():
+                if self.network.name + num == node.name:
+                    with open(node.data_dir_name / f'{Utils.EosServerName}.pid', 'r') as f:
+                        pid = int(f.readline())
+                        self.terminate_wait_pid(pid)
+
+    def start_all(self):
+        if self.args.launch.lower() != 'none':
+            for instance in self.network.nodes.values():
+                self.launch(instance)
+                time.sleep(self.args.delay)
+        self.write_bios_boot()
+
+    def write_bios_boot(self):
+        templatePath = Path(self.args.config_dir) / 'launcher' / self.args.template
+        with open(templatePath, 'r') as f:
+            template = f.read()
+        scriptPath = Path('.') / self.args.script
+        with open(scriptPath, 'w') as f:
+            bhost = self.network.nodes['bios'].host_name
+            biosport = self.network.nodes['bios'].http_port
+            prefix = '###INSERT '
+            for line in template.split('\n'):
+                if prefix in line:
+                    if 'envars' in line:
+                        f.write(f'bioshost={bhost}\n')
+                        f.write(f'biosport={biosport}\n')
+                    elif 'prodkeys' in line:
+                        for node in self.network.nodes.values():
+                            f.write(f'wcmd import -n ignition --private-key {node.keys[0].privkey}\n')
+                    elif 'cacmd' in line:
+                        for node in self.network.nodes.values():
+                            if node.producers and node.producers[0] == 'eosio':
+                                continue
+                            for producer in node.producers:
+                                f.write(f'cacmd {producer} {node.keys[0].pubkey} {node.keys[0].pubkey}\n')
```

Comment:
are both of these supposed to be `pubkey` or is one supposed to be `privkey`?

Reply:
I imitated what the C++ code was doing.  Turns out the second key was never used in the script, so I eliminated it.